File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static boolean[] visited;
    static ArrayList<Integer>[] graph;
    static int V, E, r;
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringBuilder sb = new StringBuilder();
    public static void main(String[] args) throws IOException {
        StringTokenizer st = new StringTokenizer(br.readLine());

        V = Integer.parseInt(st.nextToken());
        E = Integer.parseInt(st.nextToken());
        r = Integer.parseInt(st.nextToken());
        graph = new ArrayList[V+1];
        visited = new boolean[V+1];
        for(int i = 0; i < V + 1; i++){
            graph[i] = new ArrayList<>();
        }
        for(int i = 0; i < E; i++){
            st = new StringTokenizer(br.readLine());
            int start = Integer.parseInt(st.nextToken());
            int end = Integer.parseInt(st.nextToken());
            graph[start].add(end);
            graph[end].add(start);
        }
        for(int i= 1; i < V+1; i++){
            Collections.sort(graph[i]);
        }
        DFS(r);
        sb.append("\n");
        visited = new boolean[V+1];
        Queue<Integer> queue = new ArrayDeque<>(Arrays.asList(r));
        while(!queue.isEmpty()){
            int cur = queue.remove();
            if(!visited[cur]){
                visited[cur] = true;
                sb.append(cur + " ");
            }
            for(int d : graph[cur]){
                if(!visited[d])
                    queue.add(d);
            }
        }
        System.out.println(sb.toString());
        br.close();
    }

    static void DFS(int r){
        if(visited[r])
            return;
        visited[r] = true;
        sb.append(r + " ");
        for(int d : graph[r]){
            if(!visited[d])
                DFS(d);
        }
    }
}

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
import sys
from collections import deque
sys.setrecursionlimit(10**6)
input = sys.stdin.readline

N, M, V = map(int, input().split())

visited = [0 for _ in range(N + 1)]
graph = [[] for _ in range(N + 1)]

for i in range(M):
    s, e = map(int, input().split())
    graph[s].append(e)
    graph[e].append(s)

for i in range(1, N+1):
    graph[i].sort()

def dfs(n):
    visited[n] = 1
    print(n, end = ' ')

    for d in graph[n]:
        if visited[d] == 0:
            dfs(d)

def bfs(n):
    q = deque()
    q.append(n)

    while q:
        v = q.popleft()

        if visited[v] == 0:
            visited[v] = 1
            print(v, end=' ')

            for d in graph[v]:
                if visited[d] == 0:
                    q.append(d)

dfs(V)
print()
visited = [0 for _ in range(N + 1)]
bfs(V)